api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto

```diff
@@ -582,6 +588,35 @@ message HttpConnectionManager {
   // *not* the deprecated but similarly named :ref:`stream_error_on_invalid_http_messaging
   // <envoy_v3_api_field_config.core.v3.Http2ProtocolOptions.stream_error_on_invalid_http_messaging>`
   google.protobuf.BoolValue stream_error_on_invalid_http_message = 40;
+
+  // The configuration for the original-IP detection extensions.
+  //
+  // By default, when the :ref:`use_remote_address
+  // <envoy_api_field_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.use_remote_address>`
+  // option is not used, Envoy will rely on the :ref:`config_http_conn_man_headers_x-forwarded-for` header
+  // and the :ref:`xff_num_trusted_hops
+  // <envoy_api_field_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.xff_num_trusted_hops>`
+  // configuration option to detect the remote address.
+  //
+  // However, relying on the :ref:`config_http_conn_man_headers_x-forwarded-for`
+  // header doesn't necessarily work for all network configurations. For instance,
+  // if the number of trusted hops isn't consistent for requests flowing into
+  // Envoy this mechanism cannot be used reliably. For such cases, the following
+  // option allows for a list of custom extensions to be provided.
+  //
+  // If configured, the custom extensions will be called along with the request headers
+  // and information about the downstream connection, such as direct connected address,
+  // when the HCM needs to populate the final remote address. Each extension will then rely
+  // on these parameters to decide the value of the request's final remote address.
+  // If an extension fails to detect the original IP address and isn't configured to reject
+  // the request when this happens, the HCM will then try the remaining configured extensions
+  // until one succeeds or the request is rejected.
```

Comment:
What if none succeed?

Reply:
ah yes, we fallback to the remote address -- i'll add a note